### ISSUE TITLE
RetryPrompt only displayed once

### ISSUE THREAD

**Issue by torynfarr** [microsoft/BotFramework-Emulator#1684]
## Version
4.5.1

## Describe the bug
With the Bot Framework Emulator version 4.5.1, RetryPrompts in Dialogs are no longer working as expected. Specifically, with a text prompt, the validation method is called... but if it returns false the RetryPrompt Activity is only displayed that first time. If you enter invalid input again and again after that first time, the validation method is called each time (as it should be), but when it returns false... nothing is output.

The exact same dialog (same code, no changes) running in Direct line / Web Chat functions properly. Problem appears to be specific to the Bot Framework Emulator.

## To Reproduce

1. Create a dialog which prompts the user for text input. 
2. When text is entered, the validation method will be reached.
3. Evaluate the text and (for the sake of testing) return false as it is invalid input.
4. Observe that the RetryPrompt is displayed as it should be.
5. Enter invalid input again and the validation method will be reached again.
6. Observe that the RetryPrompt is not displayed, although it should be.
[bug]

**Comment by cwhitten** [comment]
@torynfarr we can't reproduce this. were you able to resolve your issue?

**Comment by sgellock** [comment]
closing to do lack or activity.  @torynfarr if you are still experiencing this issue, please open a new issue and we'll help

**Comment by anusharavindrar** [comment]
While testing the multi-turn-prompt C# sample on the emulator v4.5.2, I found that the retry prompt isn't showing up when I enter an invalid value for the age the second time. Instead, it waits for me to enter the value within the accepted bounds and then proceeds with the next steps. The retry prompt is displayed only once. However, when I test it on the lower version(v4.3.3), it works just fine and the retry prompt shows up every time an invalid value is entered.

Attaching the screenshots for both the versions below:

- For Emulator v4.5.2

![emulator issue](https://user-images.githubusercontent.com/48731427/65556479-9bd75100-dee4-11e9-89e7-f50275e1a653.png)

- For Emulator v4.3.3
![emulator issue 4 3 3](https://user-images.githubusercontent.com/48731427/65556498-a72a7c80-dee4-11e9-91c7-c89bfbbca9c8.png)

**Comment by cwhitten** [comment]
Hi @anusharavindrar. Thank you for re-opening this. If you look at the log panel in your screenshots, the Activity holding the invalid value message is getting sent in to the Emulator in both versions correctly, and you can inspect it normally. What isn't showing up is the message in the chat control. This is a bug and something we will fix.

Cheers

**Comment by v-kydela** [comment]
My suspicion is this:

1. The retry prompt is first set to an activity with no ID, generated by `MessageFactory.Text`.
2. Once the retry prompt gets sent, the Emulator generates an ID and returns that in the response, which the turn context then [assigns](https://github.com/microsoft/botbuilder-dotnet/blob/master/libraries/Microsoft.Bot.Builder/TurnContext.cs#L267-L277) to the activity that's set as the retry prompt.
3. When the retry prompt gets sent again, it already has an ID. I'm guessing the Emulator decides not to generate a new ID for activities that get sent with the `id` property already populated.

**Comment by compulim** [comment]
@v-kydela Are you saying the following highlighted activities have same activity ID? For example, can you repro it in Emulator and click on the highlighted rows and check?

![image](https://user-images.githubusercontent.com/1622400/65558138-72b9bf00-deea-11e9-8619-0d2a38d7fe84.png)

If yes, Web Chat will merge them into a single entry. And we would need to fix it in Emulator.

**Comment by v-kydela** [comment]
@compulim - Yes. [This customer](https://stackoverflow.com/questions/58082537/why-does-the-retry-prompt-only-show-once) has already done that detective work.

**Comment by tonyanziano** [comment]
I just tried this on both `master` and in the latest nightly build, and I'm not able to reproduce this issue. The retry prompts have unique activity IDs.:

**master**

![image](https://user-images.githubusercontent.com/3452012/66509618-45dbdf00-ea88-11e9-8666-795467304b64.png)

**Latest nightly (4.5.2-82320)**

![image](https://user-images.githubusercontent.com/3452012/66509759-876c8a00-ea88-11e9-8a39-fb375c19e3ca.png)

@torynfarr or @anusharavindrar  can you try this scenario out on the [latest nightly version](https://github.com/microsoft/BotFramework-Emulator-nightlies/releases) and see if the issue has been resolved?

**Comment by cwhitten** [comment]
@tonyanziano are you able to reproduce in the following issue? https://github.com/microsoft/BotFramework-Emulator/issues/1888

**Comment by Aliandi** [comment]
Hey @tonyanziano, we were working on this issue and we are still able to repro it with the latest nightly version but only when using a dotnet sample. Are you using a JS one?

![imagen](https://user-images.githubusercontent.com/22912283/66510709-eabbe380-eaab-11e9-9e0e-d68e4c4f99e1.png)

**Comment by anusharavindrar** [comment]
@tonyanziano I tried out the multi-prompt C# sample on the latest nightly version and I am still able to repro the issue. Attaching the screenshots below.
![em1](https://user-images.githubusercontent.com/48731427/66511122-37db8d80-ea8b-11e9-8eb8-a2b78e789d9d.png)
![em2](https://user-images.githubusercontent.com/48731427/66511131-3ad67e00-ea8b-11e9-9d94-1bdc409dced4.png)

**Comment by tonyanziano** [comment]
Hi @Aliandi ,

Yeah, I'm using a JS sample, and it seems to work fine.

So I'm starting to believe that this is a C# issue.

**Comment by tonyanziano** [comment]
@cwhitten isn't #1888 a dupe? Either way I tried to reproduce the same scenario by entering "no" instead of a number, and the retry prompt activity still appeared every time using the JS sample.

**Comment by Aliandi** [comment]
@tonyanziano, looking [here](https://github.com/microsoft/BotFramework-Emulator/issues/1888#issuecomment-536769316) it seems like there is an issue between C# and the Emulator, as the sample seems to work just fine on other channels. 
We are going to:
1. Test this sample on another channel (slack most likely) to rule out any issue with the sample itself
2. If it works fine, debug the emulator with the JS sample to find how the ID is being assigned and compare it with how it works when we use the dotnet sample

We'll update you how that goes.

**Comment by tonyanziano** [comment]
Thanks for helping out @Aliandi . Sounds good.

**Comment by v-kydela** [comment]
@Aliandi - I had already determined the cause of this issue as shown in [this comment](https://github.com/microsoft/BotFramework-Emulator/issues/1684#issuecomment-534784473). I've now raised this as a bug in the .NET repo.

**Comment by tonyanziano** [comment]
I think that this issue depends on if & how IDs are generated by both the JS and C# runtimes, and the activity ID logic on the Emulator side.

In the Emulator, if we receive an activity that already has an ID, we default to that ID and do not generate one, if it doesn't have an ID, we generate one ourselves.

With that in mind, here are the two scenarios I'm thinking of:

1. The C# runtime generates incorrect (duplicate) activity IDs for retry prompt activities, which then get passed to the Emulator and are used.

2. The C# runtime **does not** generate any IDs for the activities, and the generation is done (incorrectly) on the Emulator side for retry prompts

In either of these scenarios, the JS runtime could either be generating the correct IDs, or in fact, no IDs are generated and the Emulator is correctly generating them.

**Comment by v-kydela** [comment]
@tonyanziano 

No SDK generates an activity ID for outgoing activities. Like a conversation ID or a user ID, an activity ID is generated by the channel and the format will vary between channels. The generation is done by the Emulator, though not incorrectly. It's the Emulator's responsibility to generate an ID because it has to.

As I explained in the bug that was just linked, the difference between the .NET SDK and the Node SDK is that the .NET SDK uses the channel-generated activity ID after sending the activity to update its own activity on the bot side, so the bot now has an activity with an ID that matches the one in the channel. In the case of a retry prompt, that ID will be saved into the dialog state and used on subsequent turns, and in the case of the Emulator, the activity will reach the Emulator already having an ID that matches an activity that's already in the Emulator.

**Comment by tonyanziano** [comment]
@v-kydela ahhhh, I see. Thanks for clearing that up.
